File: trees_and_graphs/minimum_absolute_difference_in_bst.js
/**
 * @param {TreeNode} root
 * @returns {number}
 */
var getMinimumDifferenceRecursive = function(root)
{
    let dfs = node => {
        if (!node)
        {
            return [];
        }

        let left = dfs(node.left);
        const right = dfs(node.right);

        left.push(node.val);
        left.push(...right);
        return left;
    }

    let values = dfs(root);
    let ans = Infinity;
    for (let i = 1; i < values.length; i++)
    {
        ans = Math.min(ans, values[i] - values[i - 1]);
    }

    return ans;
};

/**
 * @param {TreeNode} root
 * @returns {number}
 */
var getMinimumDifferenceIterative = function(root)
{
    let iterativeInorder = root => {
        let stack = [];
        let values = [];
        let curr = root;

        while (stack.length || curr)
        {
            if (curr)
            {
                stack.push(curr);
                curr = curr.left;
            }
            else
            {
                curr = stack.pop();
                values.push(curr.val);
                curr = curr.right;
            }
        }

        return values;
    }

    let values = iterativeInorder(root);
    let ans = Infinity;
    for (let i = 1; i < values.length; i++)
    {
        ans = Math.min(ans, values[i] - values[i - 1]);
    }

    return ans;
};

File: trees_and_graphs/minimum_absolute_difference_in_bst.py
import treenode

def getMinimumDifferenceRecursive(root: treenode.TreeNode) -> int:
    def dfs(node: treenode.TreeNode) -> list[int]:
        if not node:
            return []
        
        left = dfs(node.left)
        right = dfs(node.right)
        return left + [node.val] + right
    
    values = dfs(root)
    ans = float("inf")
    for i in range(1, len(values)):
        ans = min(ans, values[i] - values[i - 1])
    
    return ans

def getMinimumDifferenceIterative(root: treenode.TreeNode) -> int:
    def iterative_inorder(root):
        stack = []
        values = []
        curr = root

        while stack or curr:
            if curr:
                stack.append(curr)
                curr = curr.left
            else:
                curr = stack.pop()
                values.append(curr.val)
                curr = curr.right

        return values
    
    values = iterative_inorder(root)
    ans = float("inf")
    for i in range(1, len(values)):
        ans = min(ans, values[i] - values[i - 1])
    
    return ans